SO-4345: Map compare improvements

molnarlaura opened the pull request.

Commit:
SO-4345: add enum for config props

Commit:
SO-4345: add ability to configure map compare props

Commit:
SO-4345: add map compare equivalence logic

Commit:
SO-4345: add ability to compare based on compare config

Commit:
SO-4345: create constants for easier usage

Commit:
Merge remote-tracking branch 'origin/7.x' into improvement/SO-4345-configure-map-compare-props

Commit:
SO-4345: fix compare logic

Commit:
SO-4345: remove equal mappings first and then search for changed ones

Commit:
SO-4345: fix compare logic

Commit:
Merge remote-tracking branch 'origin/7.x' into improvement/SO-4345-configure-map-compare-props

Commit:
SO-4345: use source and target for checking equivalence as well

Commit:
SO-4345: add equivalence class for source mapping equivalence

Commit:
SO-4345: change compare logic

Commit:
SO-4345: fix compare logic

Commit:
Merge remote-tracking branch 'origin/7.x' into improvement/SO-4345-configure-map-compare-props

Commit:
SO-4345: remove unnecessary code

Commit:
SO-4345 Comments added.

Commit:
SO-4345 Added alternative solution.

Commit:
SO-4345: change the unchanged label to Same

Commit:
SO-4345: remove alternative logic and extract common method

Commit:
SO-4345: add priority for changekind

Commit:
SO-4345: extract common method

Commit:
SO-4345: count number of total unchanged as well, limit later

Commit:
Merge remote-tracking branch 'origin/7.x' into improvement/SO-4345-configure-map-compare-props

Commit:
SO-4345: use sets.difference() to get unique mappings

Commit:
SO-4345: set default limit to the size of sets to compare

cmark commented on core/com.b2international.snowowl.core/src/com/b2international/snowowl/core/codesystem/CodeSystemRequests.java:
```diff
@@ -105,8 +107,8 @@ public static AllCodeSystemSearchRequestBuilder prepareSearchAllCodeSystems(){
 	
-	public static ConceptMapCompareRequestBuilder prepareConceptMapCompare(ComponentURI baseConceptMapURI, ComponentURI compareConceptMapURI){
-		return new ConceptMapCompareRequestBuilder(baseConceptMapURI, compareConceptMapURI);
+	public static ConceptMapCompareRequestBuilder prepareConceptMapCompare(ComponentURI baseConceptMapURI, ComponentURI compareConceptMapURI, Set<ConceptMapCompareConfigurationProperties> selectedConfig){
```
This parameter should be an option, not a mandatory parameter. By default compare by ComponentURI (CodeSystem + Code).

cmark commented on core/com.b2international.snowowl.core/src/com/b2international/snowowl/core/compare/ConceptMapCompareChangeKind.java:
```diff
@@ -21,29 +21,35 @@
-	PRESENT("Present"),
-	MISSING("Missing"),
+	PRESENT("Present", "2"),
+	MISSING("Missing", "1"),
```
How about using the enum ordinal as priority? If that's not good, then how about using `int`  for the priority field?

cmark commented on core/com.b2international.snowowl.core/src/com/b2international/snowowl/core/compare/ConceptMapCompareConfigurationProperties.java:
```diff
@@ -0,0 +1,88 @@
+ */
+public enum ConceptMapCompareConfigurationProperties {
+
+	CODE_SYSTEM(
```
⚠️ Fix the indentation!

molnarlaura replied:
Ctrl+I does not change anything, I think the indentation is OK.

cmark replied:
No, it's not 😢 Fixed it.

cmark commented on core/com.b2international.snowowl.core/src/com/b2international/snowowl/core/compare/ConceptMapCompareResult.java:
```diff
@@ -36,27 +33,27 @@
 		
-		this.addedMembers = addedMembers.stream().limit(limit).collect(toList());
-		this.removedMembers = removedMembers.stream().limit(limit).collect(toList());
+		this.addedMembers = addedMembers.stream().collect(toList());
```
Use `List.copyOf()` if possible.

molnarlaura replied:
Done

cmark commented on core/com.b2international.snowowl.core/src/com/b2international/snowowl/core/compare/ConceptMapCompareResult.java:
```diff
@@ -36,27 +33,27 @@
-		this.addedMembers = addedMembers.stream().limit(limit).collect(toList());
-		this.removedMembers = removedMembers.stream().limit(limit).collect(toList());
+		this.addedMembers = addedMembers.stream().collect(toList());
+		this.removedMembers = removedMembers.stream().collect(toList());
```
Same here.

molnarlaura replied:
Done

cmark commented on core/com.b2international.snowowl.core/src/com/b2international/snowowl/core/compare/ConceptMapCompareResult.java:
```diff
@@ -36,27 +33,27 @@
-		Builder<ConceptMapMapping, ConceptMapMapping> limitedMembers = ImmutableMultimap.builder();
-		changedMembers.entries().stream().limit(limit).forEach(limitedMembers::put);
-		this.changedMembers = limitedMembers.build();
+		this.changedMembers = changedMembers.stream().collect(toList());
```
Same here.

molnarlaura replied:
Done

cmark commented on core/com.b2international.snowowl.core/src/com/b2international/snowowl/core/compare/ConceptMapCompareResult.java:
```diff
@@ -36,27 +33,27 @@
 		this.totalAdded = addedMembers.size();
 		this.totalRemoved = removedMembers.size();
-		this.totalChanged = changedMembers.values().size();
+		this.totalChanged = changedMembers.size()/2;
```
Magic number! Divided by two? Why (because of counting both base and compare side of the change)? Please add a comment.

molnarlaura replied:
Fixed

cmark commented on core/com.b2international.snowowl.core/src/com/b2international/snowowl/core/request/ConceptMapCompareRequest.java:
```diff
@@ -16,44 +16,48 @@
 	
 	private final ComponentURI baseConceptMapURI;
 	private final ComponentURI compareConceptMapURI;
+	private final Set<ConceptMapCompareConfigurationProperties> selectedConfig;
```
Add a `@NotEmpty` validation annotation and also a default value set in the corresponding builder.

molnarlaura replied:
Done

Commit:
SO-4345: make class final

Commit:
SO-4345: make field private

cmark commented on core/com.b2international.snowowl.core/src/com/b2international/snowowl/core/request/ConceptMapCompareRequest.java:
```diff
@@ -16,44 +16,48 @@
 * @since 7.8
 */
 public final class ConceptMapCompareRequest extends ResourceRequest<BranchContext, ConceptMapCompareResult> {
+	public static final int DEFAULT_MEMBER_COMPARE_LIMIT = 50_000;
```
Where do you use this? `5000` is set on the corresponding builder.

molnarlaura replied:
It is used in the `CompareMappingSetsHandler`'s and `SnomedConceptMapCompareHandler`'s onFinish validation.

cmark replied:
Then move it to those classes, since that is UI specific and keep the API still load 5000 changes by default.

cmark commented on core/com.b2international.snowowl.core/src/com/b2international/snowowl/core/request/ConceptMapCompareRequest.java:
```diff
@@ -16,44 +16,48 @@
+	private final Set<ConceptMapCompareConfigurationProperties> selectedConfig;
 	
 	@Min(0)
 	private int limit;
```
Are we using this now? I'm not seeing any statement nor filter that uses this value.

molnarlaura replied:
It is used when creating the `MapCompareSourceAndTargetEquivalence` instance and in the `isSourceEqual()` method.

cmark commented on core/com.b2international.snowowl.core/src/com/b2international/snowowl/core/request/ConceptMapCompareRequestBuilder.java:
```diff
@@ -30,20 +33,23 @@
 	
```
`5000` limit is present in this class.

molnarlaura replied:
This only limits the size of the result, the other one is for checking whether a set is larger than the default 50 000 limit.

cmark commented on snomed/com.b2international.snowowl.snomed.core.rest.tests/src/com/b2international/snowowl/snomed/core/rest/compare/SnomedMapTypeReferenceSetCompareTest.java:
```diff
@@ -88,15 +89,15 @@ public void compareEqualSimpleMapTypeReferenceSets() {
 		ComponentURI compareURI = createURI(rf2Id);
 		
-		ConceptMapCompareResult result = CodeSystemRequests.prepareConceptMapCompare(baseURI, compareURI)
+		ConceptMapCompareResult result = CodeSystemRequests.prepareConceptMapCompare(baseURI, compareURI, ConceptMapCompareConfigurationProperties.DEFAULT_SELECTED_PROPERTIES)
```
This can be the default value of the config field in the ConceptMap Compare API.

Commit:
SO-4345: change hashCode logic

Commit:
SO-4345: add notempty annotation

Commit:
SO-4345: add notempty annotation

Commit:
SO-4345: add limited members and total numbers to result

Commit:
SO-4345: use default compare config if there is any set

cmark committed:
Merge remote-tracking branch 'origin/7.x' into improvement/SO-4345-configure-map-compare-props

cmark committed:
SO-4345: handle limit properly on server-side ConceptMap compare

Return a single list of result items instead of kind specific mapping
change lists.

cmark committed:
SO-4345: fix compile errors...

...in ConceptMapCompareSnomedMapTypeReferenceSetTest
